File: src/main/kotlin/dev/slne/surf/essentialsold/utils/arguments/WorldEnvironmentArgument.java
package dev.slne.surf.essentialsold.utils.arguments;

import dev.jorel.commandapi.arguments.ArgumentSuggestions;
import dev.jorel.commandapi.arguments.CustomArgument;
import dev.jorel.commandapi.arguments.StringArgument;
import dev.slne.surf.essentialsold.utils.brigadier.Exceptions;
import lombok.val;
import org.bukkit.World;

import java.util.Arrays;
import java.util.concurrent.CompletableFuture;

/**
 * Custom argument for {@link World.Environment}
 */
public class WorldEnvironmentArgument extends CustomArgument<World.Environment, String> {
    /**
     * A {@link World.Environment} argument
     * @param nodeName the name of the node for this argument
     */
    public WorldEnvironmentArgument(String nodeName) {
        super(new StringArgument(nodeName), info -> {
            val input = info.currentInput();
            final World.Environment environment;
            try {
                environment = World.Environment.valueOf(input.toUpperCase());
                if (environment == World.Environment.CUSTOM) throw new IllegalArgumentException(); // throw an error if the input is CUSTOM
                return environment; // try to parse the input as a World.Environment
            } catch (IllegalArgumentException e) {
                throw CustomArgumentException.fromAdventureComponent(Exceptions.ERROR_WORLD_ENVIRONMENT_NOT_EXISTS.message(input)); // throw an error if the input is not a World.Environment
            }
        });
        replaceSuggestions(ArgumentSuggestions.stringCollectionAsync(info -> CompletableFuture.supplyAsync(() -> Arrays.stream(World.Environment.values()) // suggest all World.Environments
                .filter(environment -> environment != World.Environment.CUSTOM) // don't suggest CUSTOM
                .map(worldEnvironment -> worldEnvironment.name().toLowerCase())
                .toList())));
    }
}

File: src/main/kotlin/dev/slne/surf/essentials/command/argument/world/WorldEnvironmentArgument.kt
package dev.slne.surf.essentials.command.argument.world

import dev.jorel.commandapi.CommandTree
import dev.jorel.commandapi.arguments.Argument
import dev.jorel.commandapi.arguments.ArgumentSuggestions
import dev.jorel.commandapi.arguments.CustomArgument
import dev.jorel.commandapi.arguments.StringArgument
import dev.slne.surf.surfapi.core.api.messages.adventure.buildText
import org.bukkit.World
import org.bukkit.command.CommandSender

class WorldEnvironmentArgument(nodeName: String) :
    CustomArgument<World.Environment, String>(StringArgument(nodeName), { info ->
        World.Environment.entries.firstOrNull { it.name == info.input.uppercase() }
            ?: throw CustomArgumentException.fromAdventureComponent {
                buildText {
                    appendPrefix()
                    error("Das Welt-Umfeld wurde nicht gefunden.")
                }
            }
    }) {
    init {
        this.replaceSuggestions(
            ArgumentSuggestions.stringCollection<CommandSender> {
                World.Environment.entries.map { it.name.lowercase() }
            }
        )
    }
}

inline fun Argument<*>.worldEnvironmentArgument(
    nodeName: String,
    optional: Boolean = false,
    block: Argument<*>.() -> Unit = {}
): Argument<*> = then(
    WorldEnvironmentArgument(nodeName).setOptional(optional).apply(block)
)

inline fun CommandTree.worldEnvironmentArgument(
    nodeName: String,
    optional: Boolean = false,
    block: Argument<*>.() -> Unit = {}
): CommandTree = then(
    WorldEnvironmentArgument(nodeName).setOptional(optional).apply(block)
)